### ISSUE TITLE
SQLite store performance tests

### ISSUE THREAD

**Issue by Bulat-Ziganshin** [codex-storage/nim-codex#131]
Once SQLite store will be available, we will need to have a benchmark measuring its performance for our tasks, and ideally - comparison with the FileStore backend.

**Comment by Bulat-Ziganshin** [comment]
Testing scenarios can be complex. I propose to develop testing API first. Since ATM we never delete blocks, it just have to test reading and writing in different proportions:
- `write_blocks(N): seq[Cid]`:
  - generate N random blocks
  - optionally compute their CIDs (and return them from the function)
  - optionally write blocks to DB
- `read_blocks(seq[Cid])`
- `list_blocks(): seq[Cid]`

And auxiliary functions:
- `write_list(filename, seq[Cid])` - store Cids to the file
- `read_list(filename): seq[Cid]`
- `shuffle(var seq[auto])` - reshuflle list
- `random_extract(seq[auto], N)` - extract N random elements from the list, possible repeated
- reverse, sort, take(N) - should be in sequtils?

We can also make an advanced function, mixing writing and reading blocks in given proportion:
- read_write_blocks(init: seq[Cid], R, W): seq[Cid]` - given a database with `init` records, make R reading and W writing requests in random order and return resulting list of blocks

**Comment by Bulat-Ziganshin** [comment]
The next question is how to implement scripts. AFAIK, using Nim for scripting will mean that the entire codebase will be recompiled each time we change a script. Also, we need a system for fast experiments.

So I propose to start with very simple cmdline API: `w150` means write 150 blocks, `r` means read all blocks, and `r150` means read 150 random blocks.

**Comment by emizzle** [comment]
Discussion needed.

**Comment by dryajov** [comment]
> Testing scenarios can be complex.

I would keep things as simple as possible without overcomplicating anything. 

We already have some basic unit tests for our datastores https://github.com/status-im/nim-codex/blob/main/tests/codex/stores/, lets take this as a baseline and create a parametrizable benchmark that we can launch with a combination of parameters. 

In particular I would like to see a comparison between the FS and SQL stores. As for the actual benchmark instrumentation, we can use something extremely simple as depicted here - https://stackoverflow.com/questions/36577570/how-to-benchmark-few-lines-of-code-in-nim. 

We don't have to recompile the entire project, we can compile and run the benchmark with a simple `nim c -r benchmark/storebench.nim -s=<sie-of-block> -n<num-of-blocks> ...`

So to summarize this is what I would suggest:

- Create a parametrizable benchmark based around our existing unit tests that test each individual method invocation
- Use a simple instrumentation as in the cited stackoverflow post (https://stackoverflow.com/questions/36577570/how-to-benchmark-few-lines-of-code-in-nim)
- Integrate with the rest of the build system, but only runnable on demand (e.g. `make bench`)
- Lastly, integrate with the metrics backend by using the `histogram` collector from nim-metrics - https://github.com/status-im/nim-metrics#histogram
  - This allows us to easily collect this metrics from a running application, which should be overall useful

**Comment by dryajov** [comment]
As for the actual parametrization, we should definitely allow running with different block sizes and number of blocks as I alluded in my `nim c -r benchmark/storebench.nim -s=<sie-of-block> -n<num-of-blocks> ...` example. Specifically the total number of blocks (should be in the millions) and the different sizes of blocks (8KB, 16KB, 32KB, 64KB, etc....), should help us understand the overall throughput of the stores under difference scenarios.

**Comment by Bulat-Ziganshin** [comment]
It seems that your answer is more general than my proposal

>Create a parametrizable benchmark based around our existing unit tests that test each individual method invocation

I addressed in my proposal only this topic. Unit tests just perform single getBlocks/... operation plus checks whether it prforms in correct way. Some parts of this code can be reused, but on top of it we need to generate N unique blocks, put them into store and then retrieve them by their ids. My proposal establishes framework to corrdinate this actions

> Lastly, integrate with the metrics backend

not sure what you mean. Bencnhmark is separate code that runs getBlock/.. N times. And we just need to measure how much time it required. Do you mean gathering histograms too, so we can see distribution of time spend on each getBlock?


> `-s=<sie-of-block> -n<num-of-blocks>`

it will work for writing blocks into DB, or writing followed by reading (where we can get cached data). If we want to test reading only - we need a a bit more complex language for describing scenarios, like the one i proposed:

b65536 w1000 r500 == blocksize 64K, write 1000 blocks and then read 500 of them

and then we will need a way to store cids since reading them from DB may be pretty slow and/or retrieve data into cache

**Comment by mratsim** [comment]
Relevant: https://blog.expensify.com/2018/01/08/scaling-sqlite-to-4m-qps-on-a-single-server/
